### ISSUE TITLE
Add compiler finalizers

### ISSUE THREAD

**Issue by wouterdb** [inmanta/inmanta-core#4928]
# Problem

When writing allocators, one of the most involved aspects is making sure all resources are properly flushed back and all connections are properly closed. 

It would greatly help allocator implementation if the compiler had a finalizer mechanism. 

# Expected end result

We want to be able to register finalizers both as decorator and as callback

```python

connection = None

def get_connection():
   global connection
   if connection is None:
        connection = ...
   return connection

@finalizer
def finalize_connection():
   if connection:
      connection.close()
```
with registration
```python

connection = None
def get_connection():
  global connection 
  if not connection:
       connection = ...
       finalizer(connection.close)
   return connection
```

# Solution

1. add a @finalizer decorator (similar to @plugin) that indicates a method as a finalizer
2. when the compiler finishes, prior to export, call the finalizers (even when the compiler produced an exception),
3. make sure that if the finalizer fails, this is reported: if there is already an expection from the compiler, raise that one and log the one from the finalizer. If there was no excpetion, raise the one from the finalizer instead.
4. add documentation with an example

# Rationale

1. registration Vs decoration: we support both, as the underlying code is in fact the same
2. finalize after compile or after export: we want to encourage proper closing as soon as possible, so before export. We can, later on, add a when argument to the finalizer to have multiple finalization points.
3. finalizer exceptions don't replace the compiler exception: bit risky, we may also use the multiexception to add them. I think this approach will be least likely to break e.g. explainers. 
4. don't report to the finalizer if the compile was a success: allows the use of no-args close methods directly. We can add arguments to propagate compiler state later on by adding arguments to the annotation.

**Comment by sanderr** [comment]
Approved. We could consider a more explicit `compile_finalizer` to be more explicit on when it would be called, but we could probably achieve that from the namespace as well:
```
from inmanta import compiler

@compiler.finalizer
def finalize_connection():
    ...
```

**Comment by sanderr** [comment]
Approved. We could consider a more explicit `compile_finalizer` to be more explicit on when it would be called, but we could probably achieve that from the namespace as well:
```
from inmanta import compiler

@compiler.finalizer
def finalize_connection():
    ...
```

**Comment by edvgui** [comment]
TLDR: I am fine with this proposal.

I like this because it looks really easy to implement, and to understand for the module developer.

The things that do bother me a bit are the followings:
1. We use global variables, so we must also not forget to reset them in the `ìnmanta_reset_state` method.
2. We use global variables, so the handling of the variable holding the connection is very verbose (first access it as global, then check if it is set, otherwise set it again) and this logic will be implemented everywhere we use the finalizer.

The thing I rather had in mind was the following:  
(By the way it could totally come later, and even plug into this mechanism easily)
```python
FUNC_ARGS = typing.TypeVar("FUNC_ARGS")
FUNC_RES = typing.TypeVar("FUNC_RES")


def connection(func: typing.Callable[[FUNC_ARGS], typing.Generator[FUNC_RES, None, None]]) -> typing.Callable[[FUNC_ARGS], FUNC_RES]:
    connections: typing.Dict[FUNC_ARGS, FUNC_RES] = {}

    def finish(builder: typing.Iterator) -> None:
        try:
            next(builder)
            assert False, "Only one yield is allowed"
        except StopIteration:
            pass

    def call_or_cache(args: FUNC_ARGS) -> FUNC_RES:
        if args in connections:
            return connections[args]
        
        res_builder = iter(func(args))

        # This will build the connection
        res = next(res_builder)
        connections[args] = res

        # This will teardown the connection
        finalizer(functools.partial(finish, res_builder))

    return call_or_cache


@connection
def get_connection(username: str, password: str) -> typing.Generator[MyConnection, None, None]:
    my_conn = MyConnection(username, password)
    yield my_conn
    my_conn.close()


# Get my connection once
admin = get_connection("admin", "admin")
user = get_connection("user", "user")
admin_again = get_connection("admin", "admin")

```

What I prefer about it:
1. No global variable
2. No constraints on the arguments for the function wrapped into the finalizer
3. This is nothing new, just the same mechanism than pytest's fixtures

What is less great about it:
1. The typing needs to be done correctly, otherwise it might be very confusing to use

**Comment by sanderr** [comment]
@edvgui you might be interested in [`typing.ParamSpec`](https://docs.python.org/3/library/typing.html#typing.ParamSpec). Only Python 3.10+ though.

**Comment by wouterdb** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/core-5a7c152b04a1652024289b7b/issues/inmanta/inmanta-core/4928?planning-poker) with Zenhub @FloLey @Hugo-Inmanta @sanderr
